Add support for bed explosion in custom worlds

Doc94 opened the pull request:
Currently in craftbukkit (i make PR craftbukkit#905 but its freeze for.. dont know why) the bed explosion is handle only when the world is not natural (nether, end type) and the result for sleep is NOT_POSSIBLE_HERE.

In vanilla the config "natural" only define for this case if the player can or not sleep and save spawn point, the config bed_works define in the false case for explode the bed.

Then this PR add the config bed_works for return NOT_POSSIBLE_HERE and in the event for bed explode check for the value of bed_works.

Note: not sure if is better modify the patch for the bed events.. but i dont know how to make and not explode the repository...

Doc94 committed:
Handle bed explosion correctly

Machine-Maker commented on patches/server/0739-Handle-bed-explosion-correctly.patch:
```diff
@@ -0,0 +1,33 @@
+     private Either<Player.BedSleepingProblem, Unit> getBedResult(BlockPos blockposition, Direction enumdirection) {
+         if (!this.isSleeping() && this.isAlive()) {
+-            if (!this.level.dimensionType().natural()) {
++            if (!this.level.dimensionType().natural() || !this.level.dimensionType().bedWorks()) { // Paper add not bedWorks to NOT_POSSIBLE_HERE for handle explode
```
```suggestion
+            if (!this.level.dimensionType().natural() || !this.level.dimensionType().bedWorks()) { // Paper - check if dimension allows beds (moved from BedBlock#canSetSpawn)
```

Doc94 replied:
okay, changed :3

Machine-Maker commented on patches/server/0739-Handle-bed-explosion-correctly.patch:
```diff
@@ -0,0 +1,33 @@
+                     if (entityhuman_enumbedresult != null) {
+-                        io.papermc.paper.event.player.PlayerBedFailEnterEvent event = new io.papermc.paper.event.player.PlayerBedFailEnterEvent((org.bukkit.entity.Player) player.getBukkitEntity(), io.papermc.paper.event.player.PlayerBedFailEnterEvent.FailReason.VALUES[entityhuman_enumbedresult.ordinal()], org.bukkit.craftbukkit.block.CraftBlock.at(world, finalblockposition), entityhuman_enumbedr ...
++                        boolean bedWillExplode = entityhuman_enumbedresult == Player.BedSleepingProblem.NOT_POSSIBLE_HERE && !world.dimensionType().bedWorks(); // Paper Handle explode flag
++                        io.papermc.paper.event.player.PlayerBedFailEnterEvent event = new io.papermc.paper.event.player.PlayerBedFailEnterEvent((org.bukkit.entity.Player) player.getBukkitEntity(), io.papermc.paper.event.player.PlayerBedFailEnterEvent.FailReason.VALUES[entityhuman_enumbedresult.ordinal()], org.bukkit.craftbukkit.block.CraftBlock.at(world, finalblockposition), bedWillExplode, io.p ...
```
I don't see why this diff is needed, isnt `entityhuman_enumbedresult` already taking into account if the dimension supports beds. I thought that's what the other big diff was for.

Doc94 replied:
entityhuman_enumbedresult return NOT_POSSIBLE_HERE for two cases

- If the dimension is not natural (like nether o end)
- If the dimention has set bed in not works (added in this PR)

Then need specify the correct logic for bed explode (later in the paper event the user can change this like always) because if dimension is not natural only need affect in this case for the bed use not for the explode.

All this is based in the short-cirtuit of craftbukkit (where they move the explode logic and not updated since 1.15 aprox) and the wiki.
![image](https://user-images.githubusercontent.com/3602279/128544399-19da4a69-4ccd-4804-9299-6a9cc483d41b.png)

Machine-Maker replied:
Oh I see, the bedWorks flag is what determines the blowing up, not the `natural` thing.

This is a large-ish behavioral change that might break stuff, idk how long its been this way, but I tend to be in favor of breaking more stuff than most people so I think its fine.

Do you even need to check `entityhuman_enumbedresult` then? It should always be `NOT_POSSIBLE_HERE` if `!world.dimensionType().bedWorks()` is true.

Doc94 replied:
Its based in my PR from craftbukkit where PlayerBedEnterEvent can change the result again. then if change the result to ALLOW for example its because i think the dev want avoid the bed explode and want the player sleep in teory (they can use BlockExplosionEvent for this too)

Machine-Maker replied:
I don't think you can change the bed result to NOT_POSSIBLE_HERE from PlayerBedEnterEvent.

Doc94 replied:
Checking in Craftbukkit code is rigth.. if result its set to DENY the result is OTHER_PROBLEM then only need check the bedworks... 
Then in craftbukkit the logic if very strange...

Changed the line to only check for bedWorks.

Doc94 committed:
Apply suggest for comment

Doc94 committed:
Fix bed explosions for custom dimensions

Fix logic (thanks Machine-Maker)
Note: i keep the local var for make more clear and not make a large line...

Machine-Maker reviewed (approved):
There might be some weird behavior changes with events, but this does better represent the vanilla logic.

jpenilla commented on patches/server/0739-Handle-bed-explosion-correctly.patch:
```diff
@@ -0,0 +1,33 @@
+                     if (entityhuman_enumbedresult != null) {
+-                        io.papermc.paper.event.player.PlayerBedFailEnterEvent event = new io.papermc.paper.event.player.PlayerBedFailEnterEvent((org.bukkit.entity.Player) player.getBukkitEntity(), io.papermc.paper.event.player.PlayerBedFailEnterEvent.FailReason.VALUES[entityhuman_enumbedresult.ordinal()], org.bukkit.craftbukkit.block.CraftBlock.at(world, finalblockposition), entityhuman_enumbedr ...
++                        boolean bedWillExplode = !world.dimensionType().bedWorks(); // Paper Handle explode flag
++                        io.papermc.paper.event.player.PlayerBedFailEnterEvent event = new io.papermc.paper.event.player.PlayerBedFailEnterEvent((org.bukkit.entity.Player) player.getBukkitEntity(), io.papermc.paper.event.player.PlayerBedFailEnterEvent.FailReason.VALUES[entityhuman_enumbedresult.ordinal()], org.bukkit.craftbukkit.block.CraftBlock.at(world, finalblockposition), bedWillExplode, io.p ...
```
Missing a Paper comment on this line

jpenilla commented on patches/server/0739-Handle-bed-explosion-correctly.patch:
```diff
@@ -0,0 +1,33 @@
+                     // Paper start - PlayerBedFailEnterEvent
+                     if (entityhuman_enumbedresult != null) {
+-                        io.papermc.paper.event.player.PlayerBedFailEnterEvent event = new io.papermc.paper.event.player.PlayerBedFailEnterEvent((org.bukkit.entity.Player) player.getBukkitEntity(), io.papermc.paper.event.player.PlayerBedFailEnterEvent.FailReason.VALUES[entityhuman_enumbedresult.ordinal()], org.bukkit.craftbukkit.block.CraftBlock.at(world, finalblockposition), entityhuman_enumbedr ...
++                        boolean bedWillExplode = !world.dimensionType().bedWorks(); // Paper Handle explode flag
```
```suggestion
+                        boolean bedWillExplode = !world.dimensionType().bedWorks(); // Paper - Handle explode flag
```

jpenilla reviewed (commented):
Minor formatting things, can be handled by whoever merges this
